src/tactic/polyrith.lean

```diff
@@ -0,0 +1,585 @@
+out this conversion, updates the list of atoms with atoms from the new `poly`,
+and adds the new `poly` to the top of the list of `poly`s.
+-/
+meta def fold_function : (list expr × list poly) → expr → tactic (list expr × list poly)
+| (m, poly_list) new_exp :=
+do
+  (m', new_poly) ← poly_form_of_expr transparency.reducible m new_exp,
+  return (m', poly_list ++ [new_poly])
+
+/--
+A tactic that takes in two `expr`s, and returns a `bool`.
+The first `expr` represents a type. If the second `expr` is an
+equality of this type, it returns true. If not, it returns false.
+-/
+meta def is_eq_of_type : expr → expr → tactic bool
+| expt h_eq := (do `(@eq %%R _ _) ← infer_type h_eq, unify expt R, return tt) <|> return ff
+
+/--
+A tactic that takes in an `expr` representing a type,
+and a `list expr`. the tactic returns the sublist of those
+`expr`s that are equalities of that type.
+-/
+meta def get_equalities_of_type : expr → list expr → tactic (list expr)
+| expt l := l.mfilter (is_eq_of_type expt)
+
+/--
+The prupose of this tactic is to collect all the hypotheses
+and proof terms (specified by the user) that are equalities
+of the same type as the target. It takes in an `expr` representing
+the type, a list of expressions representing the atoms
+(typically this starts as only containing
+information about the target), a `bool` representing whether the
+user used the key word "only", and a `list pexpr` of all the
+hypotheses and proof terms selected by the user.
+
+If the key word "only" is used, it collects together only those
+hypotheses/proof terms selected by the user. If not, they are
+combined with hypotheses from the local context. We throw out
+those hypotheses that are not equalities of the given type,
+and then modify each equality such that everything has been
+moved to the left of the "=" sign.
+
+The tactic returns the names of these hypotheses (as `expr`s),
+a list of atoms updated with information from all these hypotheses,
+and a list of these hypotheses converted into `poly` objects.
+-/
+meta def parse_ctx_to_polys : expr → list expr → bool → list pexpr →
+  tactic (list expr × list expr × list poly)
+| expt m only_on hyps:=
+do
+  hyps ← hyps.mmap $ λ e, i_to_expr e,
+  hyps ← if only_on then return hyps else (++ hyps) <$> local_context,
+  eq_names ← get_equalities_of_type expt hyps,
+  eqs ← eq_names.mmap infer_type,
+  eqs_to_left ← eqs.mmap equality_to_left_side,
+  (m, poly_list) ← mfoldl (fold_function) (m, []) eqs_to_left,
+  return (eq_names, m, poly_list)
+
+
+/-!
+# Connecting with Python
+
+The following section contains code that allows lean to communicate with a python script.
+-/
+
+/--
+This tactic calls python from the command line with the args in `arg_list`.
+The output printed to the console is returned as a `string`.
+It assumes that `python3` is available on the path.
+-/
+meta def sage_output (arg_list : list string := []) : tactic json :=
+let args := ["scripts/polyrith_sage.py"] ++ arg_list in
```

Comment:
I tried running `polyrith` in an external project depending on mathlib, and I got an error which I am guessing comes from this line.  The error is
```
python3: can't open file '/[my_project_directory_path]/scripts/polyrith_sage.py': [Errno 2] No such file or directory
```
Is it possible that there is a hard-coded path here which ought to have more flexibility to find the appropriate target mathlib?

Reply:
I think we need to invoke `decl_olean` on a declaration in this file in order to locate the full path of `src/tactic/polyrith.lean`, then we can build the path relative to that.

Reply:
Fixed I think!

Reply:
Indeed, it works for me now.  Thanks!

Reply:
Thanks for catching this; I was suspicious of the relative path but couldn't construct a test case